posts/go-sucks: init (#81)
NotAShelf/notashelf.dev: apps/notashelf.dev/posts/go-sucks.md

---

> title: "Why I think Go is a Terrible Language"

Reviewer: ```suggestion
title: "Why I Think Go Is a Terrible Language"
```

> description: "Rants and Ramblings on this little known language known as Go"

Reviewer: ```suggestion
description: "Rants and ramblings on this little known language known as Go"
```

date: "2026-04-26"
keywords: ["thoughts", "programming", "software"]
---

I'll start with a disclaimer that I know will bother some of you. Maybe bother
you _a lot_, or maybe make you laugh: this is not a screed from someone who
doesn't understand Go. You wish it was, and frankly, I wish it was.
Unfortunately for me, I have written Go in the past and I have also shipped Go
to "prod" in the past. In fact, I continue to write Go when I'm too lazy or
tired to care about correctness---for reasons we'll talk about shortly. It is
also worth stating out right that while I am still in the process of converting
to more... sane languages, some of my most critical infrastructure components
have been Go for as long as I can remember. They are still Go, and they still
hold.

This post has started as a Hedgedoc document I uploaded in response to questions
asking me why I think Go is designed by people who understand language design
very well, but refuse to follow established and beloved conventions. I am still
behind this statement and throughout this post you will see exactly why I
believe this. I have read much of the spec and the proposals. I have read, or at

> least try to read, the team's stated reasoning & surrounding memos. The blog

Reviewer: ```suggestion
least tried to read, the team's stated reasoning and surrounding memos. The blog
```
ampersands are for stuff like titles and abbreviations etc, not for this context

Author: I like amperstands :(

> posts about why the things are the way they are and the such. Their defenses,

Reviewer: can we look at this again
and such what

Author: "and the such" is used like "etc." here

arguments, and more.

What follows is a case. It is _my_ case about why I think Go is a terrible*

> language. This is, put most simply, is my attempt at a precise, specific and

Reviewer: ```suggestion
language. This is, put most simply, my attempt at a precise, specific and
```
two instances of 'is' the latter is what I'm removing but you can remove the former instead

deliberately uncharitable document where the language deserves what is coming.
It is about why Go is a failure of a design philosophy dressed up and paraded
around as pragmatism.

It is my pleasure to say that this is not one for beginners who seek
reassurance. This is also not for experts who are extremely comfortable with
their language. My goal is not changing your mind, but making my case out in the
open. This is also not a hate mail towards Go, but a strongly worded opinion
essay. I intend to write something truly wholesome for Rust in the future, and

> one truly vile for Zig. For now, let's talk about Go and its design choices.

Reviewer 2: > one truly vile for Zig. 

Can't wait.

## Failure by Design

The first thing I want to talk about, and perhaps the lowest hanging fruit that

> everyone reaches for is the error handling patterns of Go. This is simply

Reviewer: ```suggestion
everyone reaches for, is the error handling patterns of Go. This is simply
```
maybe a comma here

because it really is that bad. I think the designers fully understood

> exceptions, sum types, and structured error propagation. Quite sure they not

Reviewer: ```suggestion
exceptions, sum types, and structured error propagation. Quite sure they not
```
starting a sentence with "quite sure" isn't very
hm
can we add something like "I'm" or "In fact, I'm" idk

Reviewer 3: First person is implied. Doesn't need to be said, just like "it" at the beginning of this sentence.

only understand those patterns, but also see the individual beauty of it. They
knew about Haskell's `Either`, about ML's type-safe exceptions and Java's
checked exceptions which are _flawed indeed_ but at least attempt correctness.
Instead, Go enforces explicit error returns everywhere, even when the result is
repetitive boilerplate that actively obscures control flow. Here, let me give
you a better idea. A typical Go function doing three fallible operations would
look something like this:

```go
func process(path string) error {
    f, err := os.Open(path)
    if err != nil {
        return err
    }

    data, err := io.ReadAll(f)
    if err != nil {
        return err
    }

    result, err := parse(data)
    if err != nil {
        return err
    }

    return store(result)
}
```

> The actual logic (open, read, parse and store) is four lines. FOUR. The error

Reviewer: ```suggestion
The actual logic (open, read, parse, and store) is four lines. FOUR. The error
```

plumbing, however, is _twelve_. The ratio of structural signal noise is, if my

> math is correct, 3:1 and this is called explicit. Now compare it to Rust, where

Reviewer: ```suggestion
math is correct, 3:1, and this is called explicit. Now compare it to Rust, where
```

channel blocks forever on receive or send. A closed channel panics on send.

That was a lot of `nil`s in one sentence. Phew. Anywho, the point is that those
are not rules derived from a single principle, but a collection of special cases
to memorize, and getting any one of them wrong produces a runtime panic with no
compile-time indication that anything was wrong. And I think the type system's
failures run _much_ deeper than just `nil`. Go has no sum types, no exhaustive
matching, no non-nullable types, no const generics, no immutability _by
default_. There are basic tools for making illegal states unrepresentable, and
the designers knew this. They then explicitly rejected them in favor of runtime
discipline. As much as I like
[Discipline](https://www.youtube.com/watch?v=wHxOiAz4NF8), that is no substitute
for proper language design.

The absence of sum types is the most damaging. In Rust you typically write:

```rust
enum ParseResult {
    Integer(i64),
    Float(f64),
    Text(String),
    Unknown,
}
```

The compiler forces you to handle every variant. Add a new one later, and every
match expression that _doesn't_ cover it becomes a compile error. In Go you

> write an interface, use a type switch, and every unhandled case (silently) falls

Reviewer: ```suggestion
write an interface, use a type switch, and every unhandled case falls (silently)
```
this isn't a typo but it
reads better

Reviewer 3: Current reads better.

notices.

On the same cursed note, non-nullable types don't exist. Every pointer,
interface, slice, map, channel or function can _and will_ be `nil`. You cannot
declare a variable that the compiler guarantees is never nil.

In Kotlin, `String` is non-nullable and `String?` is nullable. This is enforced
at every callsite. In rust, you use `Option<T>` and the match handles the absent
case. In go, however, the best you can do is put a comment saying "THIS
PARAMETER MUST NOT BE `nil`" and hope that someone does not pass it within the
next 3 months, leading the runtime panic. Similarly, immutability is just as
absent: there is no `const` for struct fields, no `readonly`, no way to declare
a value passed to a function must not be mutated. Go has `const` for _primitive
literals_ and that's it. Everything else is just... mutable. The alternative to
immutability guarantees is "just don't mutate things!" is once again convention,
or advice. There is no guarantee, and for a language catering to beginners this
is too heavy of a footgun to hand loaded to the users.

## Let's Talk Generics

The generics situation deserves its own section, and its own _extended_

> treatment because it is _such_ a perfect illustration of team's priorities. Go

Reviewer: ```suggestion
treatment because it is _such_ a perfect illustration of the team's priorities. Go
```

1.0 shipped sometime in 2012 without generics. Which is fine. Then, for almost a
decade, the team _explicitly_ argued that code duplication is **PERFECTLY FINE**
and was preferable to abstraction, actually. The result was `interface{}` abuse
everywhere, reflection-based helpers, and the unfortunate `go generate`
pipelines that turned code generation into a first-class workflow. The library
itself either duplicated implementations for each concrete type, or punted to

> `interface{}` and runtime assertions. If you are a user, you were told to

Reviewer: ```suggestion
`interface{}` and runtime assertions. If you were a user, you were told to
```
or are and are if this is still a thing

use site, possibly in a different package. The _canonical_ workaround is:

```go
var _ io.Writer = (*MyWriter)(nil)
```

A compile-time assertion that produces a type error if `*MyWriter` doesn't
implement `io.Writer`.

Think about what that means: the language leaves intent so implicit that
developers invented a hack: _assign a nil pointer to a blank identifier with an
explicit type_. This is just to recover the ability to state intent. That this
hack is idiomatic is more damning than any external criticism. Rust uses
explicit `impl Trait for Type` declarations. The intent is in the code. You
cannot _accidentally_ implement a trait. If you change a trait's definition, the
compiler tells you exactly which `impl` blocks need updating. The contract is in
the source, not inferred from the method set.

## But The Concurrency!

Go's concurrency model is one of its marketed features and perhaps the first to
be mentioned in response when you criticize Go. It is also where the gap between
appearance and reality is widest.

Goroutines are _genuinely_ cheap.

The "share memory by communicating" slogan gestures at CSP and the channels

> _may_ look principled, however, as if to really rub salt in the wound the

Reviewer: ```suggestion
_may_ look principled, however, as if to really rub salt in the wound, the
```

Not to mention, Go has no ownership model. Any goroutine can read or write any
shared variable at any time. The type system has no concept of thread safety
whatsoever. Rust's type system makes data races impossible by construction:
`Send` and `Sync` are automatically derived marker traits, `Mutex<T>` requires
you to acquire the lock before you can touch `T`, `Arc<T>` requires `T: Send`
before it compiles. Rust doesn't make data races hard to write, it simply makes
them impossible to compile.

Channels in Go do not enforce ownership or linearity. After you send a value on
a channel, you can still use it. Nothing in the language prevents this. You
cannot encode "send exactly once, then close" either and you cannot ensure a
channel is closed by exactly one goroutine. Goroutines have no parent and no
lifecycle tied to any scope. A function can return while the goroutines it
spawned are still running, still holding references to state that should have
been released, possibly blocked forever on a channel that will never receive.
Goroutine leaks are _trivially_ easy to produce. The standard mitigation is
`context.Context`, but passing it is optional and checking the cancellation

> signal is optional and nothing enforces either. Kotlin coroutines have lexically

Reviewer: ```suggestion
`context.Context`, but passing it is optional, checking the cancellation
signal is optional, and nothing enforces either. Kotlin coroutines have lexically
```

throughout the post, Go has a vast first-party tooling. To its credit, I think
the tooling is _good_ in a way that doesn't get enough attention. `gofmt` is
excellent, `go test` being built in is excellent. `go vet` caches a subset of
common mistakes, but the typed-nil-as-interface bug described above is not
caught by `go vet`. Goroutine leak detection is not built in. Correct mutex
usage enforcement is not built in. Those require `staticcheck`, which is
_excellent_ but is third-party and requires explicit CI integration. The
baseline static analysis for a Go project is significantly weaker than the
language's correctness story requires. Build constraints until Go 1.17 were
written as magic comments: `// +build linux amd64` and not as syntax, not as a
first-class feature, as comments the toolchain parses by convention. I am rather
biased here, but this is not what good and reliable design looks like. I draw
the line at doc comments.

This matters because tooling is the usual escape hatch offered in Go's defense.
The language does not encode the invariant, but the tool will catch it. Except
the first-party tools do not catch enough of the failures the language makes

> easy. They format the code. They run the tests. They catch some obvious

Reviewer: ```suggestion
easily. They format the code. They run the tests. They catch some obvious
```
not sure if this is what you meant

Reviewer 3: Current is fine. The meaning is: mistakes that the language makes easy... for the user to make.

as part of the language. Those are not good cornerstones for a language aiming
to be simple and powerful.

The memory model was tightened in Go 1.19 and is now more precisely specified
than it was. Without an ownership model this doesn't change the practical
situation much though, violations are catchable by the race detector if they
happen to occur during a test run, and production is where you find out
otherwise. Whether a value escapes to the heap is determined by the compiler's
escape analysis, which is a heuristic, not a contract. You cannot declare that a
value must stay on the stack or specify allocation strategy for a type. For most
programs this is an acceptable tradeoff. For latency-sensitive systems it is
not, and the debugging workflow is reading escape analysis output and profiling.
There is no language-level handle on it. Go's GC has improved dramatically and
achieves sub-millisecond pause times in common workloads, but stop-the-world
pauses still occur, and under heavy allocation pressure or with large heaps the
pause behavior becomes less predictable. For trading systems, real-time control,

> audio pipelines, anything with hard latency requirements, a GC is disqualifying

Reviewer: ```suggestion
audio pipelines, or anything with hard latency requirements, a GC is disqualified
```

Reviewer 3: Partially. First correction is ok. Second is not. It means: a GC is (a) disqualifying ("feature" for a language to have).

regardless of how good it is, because "very short pause" and "no pause" are not
the same thing. Rust has no GC. Stack allocation is the default. Heap allocation
is explicit through `Box<T>`, `Arc<T>`, `Vec<T>`. You know where every

> allocation is because, well you know, you put it there.

Reviewer: ```suggestion
allocation is because, well, you know, you put it there.
```

id get rid of `you know, `

This is, if anything, where the section should land: Go is simple only if
"simple" means fewer visible concepts on the surface. Once the program has to
explain initialization, dependency identity, static analysis, reflection, unsafe
escape hatches, allocation behavior, and latency, the complexity has not
disappeared. It has merely moved out of the type system and into the runtime,
the toolchain, the build graph, and the operational culture around the codebase.

## Closing Thoughts

None of these problems are isolated. That is the actual point. A production Go
service at scale is functions returning `(value, error)` where `error` is an
opaque interface with several undocumented concrete implementations, callers
checking `errors.Is` against known sentinels and falling through silently on
unanticipated variants, goroutines outliving the requests that spawned them and
holding references to state that should have been released, a mutex somewhere
protecting the wrong data because the association was in a comment that was
accurate when written and hasn't been updated since the refactor, a nil
interface that isn't nil propagating up three layers before producing a panic

> the stack trace makes difficult to attribute.

Reviewer: ```suggestion
which the stack trace makes difficult to attribute to.
```
or something like this

Reviewer 3: Not all words need to be said. 😅 Many are implicit.

The same pattern repeats in the supposedly simple parts of the language: package
initialization hidden in `init()`, blank imports used for side-effect
registration, module identity encoded into import paths, build constraints once
parsed from comments, reflection used where the type system cannot express the
shape of the program, race detection delegated to an optional runtime tool, and
allocation behavior exposed through compiler diagnostics rather than language
guarantees. These are not separate grievances. They are the same grievance
wearing different costumes.

Every one of these failure modes is preventable in languages that encode the
relevant invariants in their type systems. Every one of them requires discipline

> and convention to avoid in Go, and discipline and conventions fail under time

Reviewer: ```suggestion
and convention to avoid in Go, and discipline and conventions fail under time,
```

Reviewer 3: They fail under "time pressure", not "time, pressure".

pressure, team growth, and the ordinary entropy of a large codebase.

The defense of Go is that all of this is manageable. With experienced teams,

> with good code review, with staticcheck in CI, with the race detector in test

Reviewer: ```suggestion
with good code review, with staticcheck in CI, and with the race detector in test
```

runs and context used consistently, Go codebases can be safe and maintainable.
This is true. The question is why the language demands that entire investment of
infrastructure and discipline to achieve properties that Rust, Haskell, Kotlin,
and Swift provide structurally. Go was designed to solve Google's specific
operational problems: fast compilation, easy onboarding for programmers of

> widely varying experience, readable code review at massive scale just to name a

Reviewer: ```suggestion
widely varying experience, and readable code review at massive scale, just to name a
```

few. And sure, these _are_ real goals, held with clear intent. The original team
understood exactly what they were leaving out. They made deliberate tradeoffs,
consistently applied.

_The problem_, however, is that optimizing for easy onboarding means in this
case (and many others) optimizing against correctness. A language you can learn
in a week is a language that does not make wrong programs hard to write. The
ceiling on what you can prove at compile time is low by design, and as codebases
grow and teams turn over and systems become load-bearing infrastructure, that
low ceiling costs you in ways that are slow, cumulative, and difficult to
attribute directly to the language rather than to the specific code that
happened to fail.

Rust is _by no means_ a perfect language. Compile times are long enough to be a
real workflow problem on large projects. The borrow checker's learning curve is
steep and genuinely so---not artificially steep, because the concepts it
enforces are non-trivial and require building a new mental model before the
errors start making sense. Async Rust is complex in ways that affect real

> programs: [^1] the story around `async` in trait methods, `dyn` and

Reviewer: ```suggestion
programs[^1]: the story around `async` in trait methods, `dyn` and
```
not sure how your footnotes work but maybe you mean this..?

Author: FWIW footnotes come after punctuation. In Markdown it's even more important because if you add a footnote to a masked link, e.g.:

```markdown
[foo]: https://foo.com

Visit [foo] [^1] for more examples!

[^1]: Terrible website, never visit
```

some formatters/linters try to concat them like `[foo][^1]` and thus break the links. It's a good rule of thumb to always put the footnotes after punctuation. I think most citation systems do agree on this as well.
